### PR TITLE
Implement coderefs search in native go

### PR DESCRIPTION
**Pull request by atrakh**
This PR replaces the 3rd party search tool (ag) with a native go implementation to read and search for files. In regards to comparing the diff, I'd recommend reviewing this PR as if it is a  complete rewrite of all search functionality

Some known improvements from `ag`:
- Removing a 3rd party dependency that has varying performance characteristics and compatibility across our supported platforms
- Faster performance. I haven't  benchmarked on a realistic CI environment (circleci, github actions, etc) yet, but I've seen up to 5x speed improvements compared to results on test repositories . (resolves https://github.com/launchdarkly/ld-find-code-refs/issues/90)
- Include files in `.ldignore`: `ag` does not support the `!` operator (https://github.com/ggreer/the_silver_searcher/issues/1233), the native implementation does.
- Edge case: support searching files with colons in their path -- any file like this would break coderefs before because of a regular expression we used to parse `ag` output
- Test coverage: a native implementation is inherently easier to test. Coverage of new code is >90%: [files.go](https://1311-156429554-gh.circle-artifacts.com/0/tmp/circle-artifacts/coverage.html#file9) [search.go](https://1311-156429554-gh.circle-artifacts.com/0/tmp/circle-artifacts/coverage.html#file10)

TODO: 
- [x] Replace unit tests
  - [x] Follow up review of old to new unit tests to make sure we didn't miss any edge cases
- [x] Test on supported CI environments

### CODE REVIEW

File: internal/search/search.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+			ProjKey:            a.ProjKey,
+			FlagKey:            a.FlagKey,
+			Aliases:            helpers.Dedupe(append(a.Aliases, b.Aliases...)),
+		},
+	}
+}
+
+// processFiles starts goroutines to process files individually. When all files have completed processing, the references channel is closed to signal completion.
+func processFiles(files chan file, references chan ld.ReferenceHunksRep, projKey string, aliases map[string][]string, ctxLines int, delimiters string) {
+	w := new(sync.WaitGroup)
```
Comment (raymonstah) [inline internal/search/search.go:170]: the zero value is useful here:
`var w sync.WaitGroup`

File: internal/search/files.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		} else if isDir {
+			return nil
+		}
+
+		fileWg.Add(1)
+		go func() error {
+			defer fileWg.Done()
+			lines, err := readFileLines(path)
+			if err != nil {
+				return err
```
Comment (raymonstah) [inline internal/search/files.go:89]: looks like this error gets swallowed.

File: internal/search/search.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+			Lines:              strings.Join(combinedLines, "\n"),
+			ProjKey:            a.ProjKey,
+			FlagKey:            a.FlagKey,
+			Aliases:            helpers.Dedupe(append(a.Aliases, b.Aliases...)),
+		},
+	}
+}
+
+// processFiles starts goroutines to process files individually. When all files have completed processing, the references channel is closed to signal completion.
+func processFiles(files chan file, references chan ld.ReferenceHunksRep, projKey string, aliases map[string][]string, ctxLines int, delimiters string) {
```
Comment (raymonstah) [inline internal/search/search.go:169]: for safety you can specify the channel direction:
https://gobyexample.com/channel-directions

File: internal/search/search.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		return ret[i].Path < ret[j].Path
+	})
+
+	totalHunks := 0
+	for reference := range references {
+		ret = append(ret, reference)
+
+		// Reached maximum number of files with code references
+		if len(ret) >= maxFileCount {
+			return ret, nil
```
Comment (raymonstah) [inline internal/search/search.go:211]: if this returns early, what happens to the background goroutine writing into this channel?
